sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1241,6 +1242,23 @@ object OptimizeRepartition extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Remove window partition if partition expressions are foldable
+ */
+object EliminateWindowPartitions extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = removeWindowExpressionPartitions(plan)
+    .transformWithPruning(_.containsPattern(WINDOW), ruleId) {
+      case w @ Window(_, ps, _, _) if ps.exists(_.foldable) =>
+        w.copy(partitionSpec = ps.filter(!_.foldable))
+    }
+
+  def removeWindowExpressionPartitions(plan: LogicalPlan): LogicalPlan =
+    plan.transformAllExpressionsWithPruning(_.containsPattern(WINDOW_EXPRESSION), ruleId) {
+      case we @ WindowExpression(_, ws @ WindowSpecDefinition(ps, _, _)) if ps.exists(_.foldable) =>
+        we.copy(windowSpec = ws.copy(partitionSpec = ps.filter(!_.foldable)))
+    }
+}
```

Comment:
It seems we can simplify the code here with `transformWithPruning` only, then we can remove the `removeWindowExpressionPartitions`.

Reply:
If  do as you say, this will cause the `partitionSpec` in `WindowExpression` to be inconsistent with the `partitionSpec` in `Window`, which may cause hidden dangers.

Reply:
I mean is merge the logic of `removeWindowExpressionPartitions` into `transformWithPruning`.

Reply:
Ok, done.